src/hmem_rocr.c

```diff
@@ -240,6 +394,160 @@ int rocr_copy_to_dev(uint64_t device, void *dest, const void *src,
 	return ret;
 }
 
+static int
+rocr_dev_async_copy(void *dst, const void *src, size_t size,
+		    void **stream)
+{
+	void *src_hsa_ptr;
+	void *dst_hsa_ptr;
+	int ret, i;
+	hsa_status_t hsa_ret;
+	struct ofi_hsa_stream *s;
+	struct ofi_hsa_signal_info *ipc_signal;
+	/* 0 - source agent
+	 * 1 - destination agent
+	 */
+	hsa_agent_t agents[2];
+	bool src_local, dst_local;
+
+	if (!stream)
+		return -FI_EINVAL;
+
+	ret = rocr_host_memory_ptr((void *)src, &src_hsa_ptr, &agents[0],
+				   NULL, NULL, &src_local);
+	if (ret != FI_SUCCESS)
+		return ret;
+
+	ret = rocr_host_memory_ptr(dst, &dst_hsa_ptr, &agents[1], NULL, NULL,
+				   &dst_local);
+	if (ret != FI_SUCCESS)
+		return ret;
+
+	if (!*stream) {
+		pthread_spin_lock(&fs_lock);
+		s = ofi_freestack_pop(ipc_stream_fs);
+		pthread_spin_unlock(&fs_lock);
+		if (!s)
+			return -FI_ENOMEM;
+		memset(s, 0, sizeof(*s));
+	} else {
+		s = *stream;
+	}
+
+	pthread_spin_lock(&fs_lock);
+	s->sinfo[s->num_signals] = ofi_freestack_pop(ipc_signal_fs);
+	pthread_spin_unlock(&fs_lock);
+	ipc_signal = s->sinfo[s->num_signals];
+
+	s->num_signals++;
+
+	/* device to device */
+	if (!src_local && !dst_local) {
+		hsa_ret = ofi_hsa_amd_agents_allow_access(2, agents, NULL,
+							  dst_hsa_ptr);
+		if (hsa_ret != HSA_STATUS_SUCCESS) {
+			FI_WARN(&core_prov, FI_LOG_CORE,
+			   "Failed to perform hsa_amd_agents_allow_access %s\n",
+			   ofi_hsa_status_to_string(hsa_ret));
+			ret = -FI_EINVAL;
+			goto fail;
+		}
+		ipc_signal->addr = NULL;
+	/* device to host */
+	} else if (!src_local && dst_local) {
+		size_t d2h_thresh;
+
+		if (fi_param_get_size_t(&core_prov, "rocr_d2h_threshold",
+					&d2h_thresh) < 0)
+			d2h_thresh = D2H_THRESHOLD;
+		if (size < d2h_thresh) {
+			memcpy(dst, src, size);
+			ofi_hsa_signal_store_screlease(ipc_signal->sig, 0);
+			ipc_signal->addr = NULL;
+			goto finish;
+		}
+		hsa_ret = ofi_hsa_amd_memory_lock(dst, size, NULL, 0,
+						  &dst_hsa_ptr);
+		if (hsa_ret != HSA_STATUS_SUCCESS) {
+			ret = -FI_EINVAL;
+			goto fail;
+		}
+		ipc_signal->addr = dst;
+		agents[1] = agents[0];
+	}
+
+	ofi_hsa_signal_store_screlease(ipc_signal->sig, 1);
+
+	hsa_ret = ofi_hsa_amd_memory_async_copy(dst_hsa_ptr, agents[1],
+				src_hsa_ptr, agents[0],
+				size, 0, NULL, ipc_signal->sig);
```

Comment:
The "event" associated with the async copy is ipc_signal->sig.  However, that's not what the 'stream' parameter returns.  'stream' is pointing to an array of these signals.  So, a single stream parameter actually references an arbitrary number of async copies.  This shows up as the for loop in the query call below.  This is why I'm saying that the behavior is cuda-like -- query only succeeds of all the copies have completed.

I think the general flow here is good.  But we should expose APIs for the caller to allocate and free whatever this 'event/barrier/signal/async_cntr/thingy' is, along with some typedef to enable type checking.  That's much easier for a developer to figure out how to use than having some void ** parameter act as input sometimes and output other times.

If we allow multiple async operations to be associated with this, that seems reasonable.  I don't think the changes would be too significant from what's here.